packages/core/src/interpreter.ts

```diff
@@ -409,16 +406,33 @@ export class Interpreter<
             State.from(initialState, this.machine.context)
           );
 
-    this.scheduler.initialize(() => {
-      this.update(resolvedState, initEvent as SCXML.Event<TEvent>);
+    this._state = resolvedState;
+
+    this.update(resolvedState);
+
+    if (this.options.devTools) {
+      this.attachDevTools();
+    }
+
+    this.flush();
 
-      if (this.options.devTools) {
-        this.attachDevTools();
-      }
-    });
     return this;
   }
 
+  private flush() {
+    this.mailboxStatus = 'processing';
+    while (this.mailbox.length) {
+      const event = this.mailbox.shift()!;
```

Comment:
I wonder if we should implement a better data structure for the mailbox, shifting arrays is not the best for performance

Reply:
Using a linked list would make dequeuing O(1) instead of O(n), at the cost of more memory 🤔

Reply:
Added to #2870